extensions/amp-story/1.0/amp-story.js

```diff
@@ -2237,13 +2245,38 @@ export class AmpStory extends AMP.BaseElement {
 
     const pagesByDistance = this.getPagesByDistance_();
 
-    this.mutateElement(() => {
-      pagesByDistance.forEach((pageIds, distance) => {
+    if (!isExperimentOn(this.win, 'amp-story-load-first-only')) {
+      return pagesByDistance.forEach((pageIds, distance) => {
         pageIds.forEach((pageId) => {
           const page = this.getPageById(pageId);
           page.setDistance(distance);
         });
       });
+    }
+
+    const preloadAllPages = (overrideDistances) => {
+      pagesByDistance.forEach((pageIds, distance) => {
+        pageIds.forEach((pageId) => {
+          const page = this.getPageById(pageId);
+          if (overrideDistances || !page.hasDistance()) {
+            page.setDistance(distance);
+          }
+        });
+      });
+    };
+
+    this.mutateElement(() => {
+      if (prioritizeActivePage) {
+        // Load page with distance 0 first, and then load the other ones.
+        return Promise.all(
+          pagesByDistance[0].map((pageId) => {
+            const page = this.getPageById(pageId);
+            page.setDistance(0);
+            return page.signals().whenSignal(CommonSignals.LOAD_END);
+          })
+        ).then(() => preloadAllPages(/* overrideDistances */ false));
```

Comment:
this new parameter with `overrideDistances` is way too complicated/hard to understand without context/easy to break.
Please avoid tying the implementation to some business logic when it's not needed. Use a deferred and cancel it when needed so your success callback doesn't run.

Reply:
Using a check for `getDistance() == 0` to know if the preloadAllPages should be called, since I figured is an easier way to check for navigation that doesn't rely on listeners or defers. LMK if this works.